syntax = "proto3";

message Wrapper {
    oneof MessageType {
        ChangeAngle changeAngle = 1;
        ChangeDir changeDir = 2;
        DoDash doDash = 3;
        ChangeBulletState changeBulletState = 4;
        GameState gameState = 5;
        GameCount gameCount = 6;
        GameInit gameInit = 7;
        GameStart gameStart = 8;
        GameReconnect gameReconnect = 9;
    }
}

// server → client : init message
message GameInit {
    int32 id = 1;  // player id
    int32 minPlayerNum = 2;
}

// client → server : changing state
message ChangeDir {
    float angle = 1;
    bool isMoved = 2;
}

message ChangeAngle {
    float angle = 1;
}

message GameReconnect {
    int32 id = 1;
    int32 minPlayerNum = 2;
    int32 totalPlayerNum = 3; // null일 수 있음
}

// server → client
message GameState {
    repeated PlayerState playerState = 1;
    repeated BulletState bulletState = 2;
    repeated HealPackState healPackState = 3;
    repeated MagicItemState magicItemState = 4;
    repeated TileState tileState = 5;
    int32 lastSec = 6;
    ChangingState changingState = 7;
}

message PlayerState {
    int32 id = 1;
    float x = 2;
    float y = 3;
    int32 health = 4;
    int32 magicType = 5;
    float angle = 6;
    int32 dashCoolTime = 7;
    bool isMoved = 8;
    bool isDashing = 9;
    bool isBurn = 10;
    bool isShooting = 11;
}

message ChangeBulletState {
    bool isShooting = 1;
}

message BulletState {
    int32 bulletId = 1;
    int32 bulletType = 2;
    // 0: snow
    // 1: stone
    // 2: fire

    float x = 3;
    float y = 4;
}

message GameCount {
    int32 count = 1;
}

message DoDash {
    bool dash = 1;
}

message GameStart {
    int32 mapLength = 1;
    int32 totalPlayerNum = 2;
}

message HealPackState {
    int32 itemId = 1;
    float x = 2;
    float y = 3;
}

message MagicItemState {
    int32 itemId = 1;
    int32 magicType = 2;
    float x = 3;
    float y = 4;
}

message TileState {
    int32 tileId = 1;
    float x = 2;
    float y = 3;
    int32 tileType = 4;
    int32 tileState = 5;
    // 0: tile normal
    // 1: tile dangerous (tile fall in 5 sec)
}

message ChangingState {
    repeated HitBulletState hitBulletState = 1;
    repeated GetItemState getItemState = 2;
    repeated PlayerDeadState playerDeadState = 3;
}

message HitBulletState {
    int32 objectType = 1;
    // 0: player
    // 4: environment object

    int32 bulletId = 2;
    int32 objectId = 3;
    float x = 4;
    float y = 5;
    int32 bulletType = 6;
    // 0: snow
    // 1: stone
    // 2: fire
}

message GetItemState {
    int32 itemId = 1;
    int32 playerId = 2;
    int32 itemType = 3;
    // 1: healpack
    // 2: stone magic
    // 3: fire magic

    float x = 4;
    float y = 5;
}

message PlayerDeadState {
    int32 killerId = 1;
    int32 deadId = 2;
    int32 dyingStatus = 3;
    // 1: snow
    // 2: stone
    // 3: fire
    // 4: fall

    int32 killNum = 4;
    int32 Placement = 5;
}